tests/unit/test_endpoint_provider.py

```diff
@@ -615,3 +615,26 @@ def test_unit_label_is_retained_if_hard_coded(self):
             for rule in group["rules"]:
                 self.assertIn("juju_unit", rule["labels"])
                 self.assertIn("juju_unit=", rule["expr"])
+
+
+class TestNoLeader(unittest.TestCase):
+    """Tests the case where leader is not set immediately."""
+
+    def setUp(self):
+        self.harness = Harness(EndpointProviderCharm, meta=PROVIDER_META)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.set_leader(False)
+        self.harness.begin_with_initial_hooks()
+
+    @patch("ops.testing._TestingModelBackend.network_get")
+    def test_alert_rules(self, _):
+        """Verify alert rules are added when leader is elected after the relation is created."""
+        rel_id = self.harness.add_relation(RELATION_NAME, "provider")
+        self.harness.add_relation_unit(rel_id, "provider/0")
+        self.harness.set_leader(True)
```

Comment:
```suggestion
        # First, make sure there are no rules present
        data = self.harness.get_relation_data(rel_id, self.harness.model.app.name).get(
            "alert_rules"
        )
        self.assertIsNone(data)
        
        # Now, trigger leader elected and make sure the rules are in place
        self.harness.set_leader(True)
```

Reply:
At first I thought this was a good idea but when I thought about it more I think we shouldn't check for this. It would be checking an implementation detail instead of checking the intended behavior. In other words, It wouldn't be a problem if alert rules showed up before `leader_elected`.

Reply:
Yep, makes sense.